pkg/sources/source_manager.go

```diff
@@ -283,25 +242,36 @@ func (s *SourceManager) runWithoutUnits(ctx context.Context, handle handle, sour
 	defer wg.Wait()
 	defer close(ch)
 	if err := source.Chunks(ctx, ch); err != nil {
-		report.AddError(err)
-		return report, err
+		report.ReportError(Fatal{err})
+		return Fatal{err}
 	}
-	return report, nil
+	return nil
 }
 
 // runWithUnits is a helper method to run a Source that is also a
 // SourceUnitEnumChunker. This allows better introspection of what is getting
 // scanned and any errors encountered.
-func (s *SourceManager) runWithUnits(ctx context.Context, handle handle, source SourceUnitEnumChunker, report *JobReport) (*JobReport, error) {
-	reporter := &mgrUnitReporter{
+func (s *SourceManager) runWithUnits(ctx context.Context, handle handle, source SourceUnitEnumChunker, report *JobReport) error {
+	unitReporter := &mgrUnitReporter{
 		unitCh: make(chan SourceUnit),
+		report: report,
+	}
+	fatalErr := make(chan error, 1)
+	catchFirstFatal := func(err error) {
+		select {
+		case fatalErr <- err:
+		default:
+		}
```

Comment:
Attempts to write the `err` into the `fatalErr` channel, throwing away all but the first one to make it (since `fatalErr` has a buffer size of 1).

I'll add a comment to make it clearer.